### PR TITLE
L025 should look at subqueries

### LINKED ISSUE #1462: L025 removes alias which is used in subquery
_If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant._

## Expected Behaviour
With fixing ~L031~ L025 the alias should replaced correctly also in subqueries.

## Observed Behaviour
A alias from the parent query is used in a subquery. After fixing the file, the alias of the parent was removed while the alias within the subquery remains.

## Steps to Reproduce
Fix this script:
```
WITH base AS (
    SELECT
        nvl(col1, col2) AS col1,
        col2,
        col3,
        col4,
        col5
    FROM
        SRC_PRD.MDM_EXTRACT
),
without_dup AS (
    SELECT
        DISTINCT col1,
        col2,
        col3,
        col4
    FROM
        base
)
SELECT
    col1,
    col2,
    col3,
    col4,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'X'
    ) AS X_CNT,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'Y'
    ) AS Y_CNT,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'Z'
    ) AS Z_CNT,
FROM
    without_dup a;
```

Alias `a` from `without_dup` is removed but lives in the where clause of the subselects.

## Dialect
exasol

## Version
Include the output of `sqlfluff --version` along with your Python version
sqlfluff, version 0.6.6
Python 3.8.12
## Configuration
default

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made

Fixes #1462

This is basically a rewrite of `L025`. Previously, it inherited from `L020`. Now, its implementation largely relies on `SelectCrawler`. A couple of other rules needlessly inherited from `L025` that could've simply inherited from `L020` -- I updated those, because after the rewrite, inheriting from `L025` no longer worked.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/rules/L025.py [hunk 1]
```diff
@@ -38,50 +59,78 @@ class Rule_L025(Rule_L020):
...
+            cls._analyze_table_aliases(cast(L025Query, child), dialect)
+
+    @classmethod
+    def _resolve_and_mark_reference(cls, query: L025Query, ref: str):
+        # Does this query define the referenced alias?
+        if any(ref == a.ref_str for a in query.aliases):
+            # Yes. Record the reference.
+            query.tbl_refs.add(ref)
+        elif query.parent:
+            # No. Recursively check the query's parent hierarchy.
```
Comment (barrywhart) [inline src/sqlfluff/rules/L025.py:113]: This is the key area of the fix: Now we'll correctly resolve a reference to a table alias that was defined in an "ancestor" query.

File: src/sqlfluff/rules/L025.py [hunk 2]
```diff
@@ -38,50 +59,78 @@ class Rule_L025(Rule_L020):
...
+            cls._resolve_and_mark_reference(cast(L025Query, query.parent), ref)
+
+    @classmethod
+    def _report_unused_alias(cls, alias: AliasInfo) -> LintResult:
```
Comment (barrywhart) [inline src/sqlfluff/rules/L025.py:117]: This code is identical to lines 66-86 of the old implementation, so it should not require a close review.

**Comment by tunetheweb** [comment]
@barrywhart does this also solve #2251 and #2030 ?

**Comment by barrywhart** [comment]
It does not address #2251. That appears to be related to an unusual alias syntax supported by the Spark3 dialect. I just commented on that other issue to ask for more context.

It also does not address #2030. That issue, also, involves something dialect-specific -- in that case, BigQuery. This reminds me a bit of some previous challenges we've had with BigQuery, where its support of structures makes it difficult for SQLFluff to understand the query.

It would be good to fix those at some point. Lately, I've been mostly focused on issues where the user's SQL is corrupted or broken, and I've prioritizing things approximately as follows:
* Fix issues that cut across multiple rules (e.g. template-related corruption).
* Fix rule issues that affect the ANSI (or most) dialects.
* Rule-specific, dialect-specific issues: I have not prioritized these due to time constraints. Would be cool to get more people working on rules. 💪

File: src/sqlfluff/rules/L025.py [hunk 3]
```diff
@@ -38,50 +59,78 @@ class Rule_L025(Rule_L020):
...
+                # Record the aliases.
+                query.aliases += select_info.table_aliases
+
+                # Look at each table reference; if its an alias reference,
+                # resolve the alias: could be an alias defined in "query"
+                # itself or an "ancestor" query.
+                for r in select_info.reference_buffer:
+                    for tr in r.extract_possible_references(level=r.ObjectReferenceLevel.TABLE):  # type: ignore
+                        # This function walks up the query's parent stack if necessary.
+                        # Reuses the existing "for r in references:" loop from L025.
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L025.py:99]: Should we remove this line?

```suggestion
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]
This looks quite nice. Need to find some time to dig into your rule work more.

One suggestion to remove a comment.

Also any performance difference (either better or worse) with this re-implementation? Any easy way of benchmarking that?

File: src/sqlfluff/rules/L025.py (hunk 3 above)
Comment (barrywhart) [reply src/sqlfluff/rules/L025.py:99]: Oh yes -- that's left over from my pseudocode.

**Review by tunetheweb — APPROVED** [review]
LGTM. Would still be nice to know if it's more or less performant but since we don't demand that for other PRs I'll not hold up approval for that.

**Comment by barrywhart** [comment]
In the typical "nothing to do" case (where it's not the target segment type or the query defines no aliases), performance should be similar. When there _are_ aliases, it'll be slower, because now it's doing a recursive walk of the parse tree underneath in order to find subqueries. The `SelectCrawler` class, which does the work, is already used by L044 and L045, so I wouldn't expect any big surprises here.

At some point, if enough rules use `SelectCrawler` and we are worried about performance, we could look at ways to cache/reuse this "query tree" info across rules (i.e. if several rules want a `SelectCrawler` on the same query, avoid having each rule redo the work).

**Comment by barrywhart** [comment]
I don't know if we have a typical way to benchmark performance. It would be easy to get precise timings from each of the rule test cases, but those are usually much smaller than real SQL. It could be cool if there were some real-world open-source SQL projects we could monitor how SQLFluff performs with.

**Comment by barrywhart** [comment]
@tunetheweb: I removed the extraneous comment and answered your questions. Ready for another review when you have time. Thanks!

**Pull request merged**